compiler/src/dotty/tools/dotc/core/Decorators.scala

```diff
@@ -131,7 +131,7 @@ object Decorators {
     }
 
     /** Union on lists seen as sets */
-    def | (ys: List[T]): List[T] = xs ++ (ys filterNot (xs contains _))
+    def | (ys: List[T]): List[T] = xs ::: (ys filterNot (xs contains _))
```

Comment:
Maybe we should do `xs filterNot (ys contains _) ::: ys` to always share the tail.

Reply:
Maybe. I did't want to change the semantic. Maybe ordering matters.

I suggest trying out optimisations in another PR. Definitely worth investigating!

Reply:
Agreed